scio-core/src/main/scala/com/spotify/scio/coders/Coder.scala

```diff
@@ -318,105 +248,249 @@ final private[scio] case class RecordCoder[T](
+    bcoder,
+    toHash: java.lang.Integer,
+    fromHash: java.lang.Integer
+  )
+  override def encode(value: T, os: OutputStream): Unit =
+    bcoder.encode(to(value), os)
+
+  override def encode(value: T, os: OutputStream, context: BCoder.Context): Unit =
+    bcoder.encode(to(value), os, context)
+
+  override def decode(is: InputStream): T =
+    from(bcoder.decode(is))
+
+  override def decode(is: InputStream, context: BCoder.Context): T =
+    from(bcoder.decode(is, context))
+
+  override def verifyDeterministic(): Unit =
+    bcoder.verifyDeterministic()
+
+  // Here we make the assumption that mapping functions are idempotent
+  override def consistentWithEquals(): Boolean =
+    bcoder.consistentWithEquals()
+
+  override def structuralValue(value: T): AnyRef =
+    bcoder.structuralValue(to(value))
+
+  override def isRegisterByteSizeObserverCheap(value: T): Boolean =
+    bcoder.isRegisterByteSizeObserverCheap(to(value))
+
+  override def registerByteSizeObserver(value: T, observer: ElementByteSizeObserver): Unit =
+    bcoder.registerByteSizeObserver(to(value), observer)
+}
+
+sealed abstract private[scio] class WrappedCoder[T] extends StructuredCoder[T] {
+  def bcoder: BCoder[T]
+
+  override def getCoderArguments: JList[_ <: BCoder[_]] =
+    Collections.singletonList(bcoder)
+
+  override def encode(value: T, os: OutputStream): Unit =
+    bcoder.encode(value, os)
+  override def encode(value: T, os: OutputStream, context: BCoder.Context): Unit =
+    bcoder.encode(value, os, context)
+  override def decode(is: InputStream): T =
+    bcoder.decode(is)
+  override def decode(is: InputStream, context: BCoder.Context): T =
+    bcoder.decode(is, context)
+  override def verifyDeterministic(): Unit =
+    bcoder.verifyDeterministic()
+  override def consistentWithEquals(): Boolean =
+    bcoder.consistentWithEquals()
+  override def structuralValue(value: T): AnyRef =
+    bcoder.structuralValue(value)
+  override def isRegisterByteSizeObserverCheap(value: T): Boolean =
+    bcoder.isRegisterByteSizeObserverCheap(value)
+  override def registerByteSizeObserver(value: T, observer: ElementByteSizeObserver): Unit =
+    bcoder.registerByteSizeObserver(value, observer)
+}
+
+final private[scio] class RefCoder[T](var bcoder: BCoder[T]) extends WrappedCoder[T] {
+  def this() = this(null)
+  override def toString: String = bcoder.toString
+}
+
+final private[scio] class LazyCoder[T](val typeName: String, bc: => BCoder[T])
+    extends WrappedCoder[T] {
+
+  override lazy val bcoder: BCoder[T] = bc
+
+  override def toString: String = s"LazyCoder[$typeName]"
+
+  // stop call stack and only compare on typeName
+  override def equals(obj: Any): Boolean = obj match {
+    case that: LazyCoder[_] => typeName == that.typeName
+    case _                  => false
+  }
+
+  override def hashCode(): Int = typeName.hashCode
+
+  // stop call stack and not interfere with other result
+  override def verifyDeterministic(): Unit = {}
```

Comment:
I'm trying to think about possible ramifications for not making the `verifyDeterministic` call and whether that could cause issues.  I wonder if we could invoke it when `bCoder` is initialized, like:

```scala
override lazy val bcoder: BCoder[T] = {
  verifyDeterministic()
  bc
}
```

wdyt?

Reply:
This is not an initialization problem. `verifyDeterministic` on a structured coder will call recursively on every field, if we reach a `LazyCoder`, we do not want to call again on the base struct. we just stop here. It is fine to not throw. determisistic decision is left to the other fields

Reply:
ah right, of course 👍